### ISSUE TITLE
Segfault in redisReaderFeed when the machine is OOM.

### ISSUE THREAD

**Issue by dpayne** [redis/hiredis#300]
A segfault occurrs when inserting large amounts of keys to a redis server when the machine has no memory left. Instead of failing gracefully, hiredis segfaults in redisReaderFeed. 

I've made a potential fix here, https://github.com/dpayne/hiredis/commit/81bb5459a08442fc4070fae334ff7c30ed742517

**Comment by badboy** [comment]
I don't see how the proposed fix will prevent a segfault. A few lines above is already used and later asserted to not be null. 
In which case will `r->buf` be `NULL` _after_ the `if` above?

**Comment by dpayne** [comment]
It's been a long time since I've looked at this code and I've since stopped using redis so I don't remember the exact details. From what I remember, the call to sdscatlen threw a segfault when the machine was out-of-memory. Since I was compiling the library from scratch and linking it with another code base, it's possible asserts were turned off. Sorry for the lack of details, this was quite a long time ago and I've since stopped using redis in the project.
